### PR TITLE
Add awsfargate module for AWS ECS Fargate

### PR DESCRIPTION
**Pull request by kaiyan-sheng**
## What does this PR do?

This PR is to add `awsfargate` module to collect monitoring metrics from AWS ECS Fargate task management endpoint for each container running under the same task as the Metricbeat is running. 

Note: `~internal~ecs~pause` container is skipped.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

I created an AWS [CloudFormation](https://github.com/kaiyan-sheng/beats/blob/fargate_module/x-pack/metricbeat/module/awsfargate/cloudformation.yml) file to create a cluster, define IAM role, create a task definition, and start the actual service. 

Create an Elastic cloud env for testing and store the credentials into AWS secret manager.
```
aws --region us-east-2 ecretsmanager create-secret --name ELASTIC_CLOUD_AUTH --secret-string elastic:XXX
```
and 
```
aws --region us-east-2 ecretsmanager create-secret --name ELASTIC_CLOUD_ID --secret-string YYYY
```
After storing the credentials into AWS secret manager,  copy[ this file](https://github.com/kaiyan-sheng/beats/blob/fargate_module/x-pack/metricbeat/module/awsfargate/cloudformation.yml) locally and copy the ARNs of these two secrets and put them into AWS CloudFormation yml file under `TaskDefinition` -> `ContainerDefinition`.
```
aws --region us-east-2 cloudformation create-stack --stack-name test-metricbeat-deployment --template-body file://./module/awsfargate/cloudformation.yml --parameters 'ParameterKey=SubnetID,ParameterValue=<subnet-id>' --capabilities CAPABILITY_NAMED_IAM
```

Make sure to replace`<subnet-id>` with your own subnet in this command. Please go to `Services -> VPC -> Subnets` to find subnet ID to use. You can also add several more containers under the `TaskDefinition` section.

Then you should see your task definition and cluster getting created, and containers start running in a minute or two. 

Make sure there is no error log for the container and then go to Elastic Cloud to check for events from this module.

Clean up after test:
```
aws cloudformation delete-stack --stack-name test-metricbeat-deployment
```

## Sample Event Output

[data.json](https://github.com/kaiyan-sheng/beats/blob/fargate_module/x-pack/metricbeat/module/awsfargate/task_stats/_meta/data.json) 

## Screenshots
Screen Shot 2020-10-21 at 7 21 23 PM

closes https://github.com/elastic/beats/issues/22210

### CODE REVIEW

**Comment by sorantis** [comment]
@kaiyan-sheng any reason this is a standalone module and note another metricset in `aws` module?

**Comment by kaiyan-sheng** [comment]
@sorantis `awsfargate` does not require any AWS credentials setup in the module configuration because it can only be run inside a fargate task in order to have access to `ECS_CONTAINER_METADATA_URI_V4 ` ECS fargate task metadata endpoint. If we create a `fargate` metricset under current `aws` module, that means it requires the AWS credentials as a part of its configuration even though it's not gonna be used.

**Comment by sorantis** [comment]
thanks for clarifying @kaiyan-sheng. Would this also translate to a standalone package?

**Comment by kaiyan-sheng** [comment]
@sorantis I'm not sure on the package side. If it's one module map to one package, then this will be a standalone package. Do we have any package include more than one module?

**Comment by sorantis** [comment]
i think it'll result in a separate package simply because it requires different configuration.

**Comment by ChrsMark** [comment]
> i think it'll result in a separate package simply because it requires different configuration.

👍

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		taskMeta := taskStats.(map[string]interface{})
+		readTimestamp := taskMeta["read"].(string)
+		timestamp, err := time.Parse(time.RFC3339, readTimestamp)
+		if err != nil {
+			m.logger.Warn(fmt.Errorf("parsing timestamp %s failed, use current timestamp in event instead", readTimestamp))
+		} else {
+			event.Timestamp = timestamp
+		}
+
+		cpuStats := taskMeta["cpu_stats"].(map[string]interface{})
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:109]: Not sure if applicable but maybe we can use a `Schema` here like waht we do for https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/metricbeat/module/nats/stats/data.go#L42

Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go]: Yep! I just added all schemas into data.go file.

**Comment by kaiyan-sheng** [comment]
TODO/Questions:

- [ ] Add documentation
- [ ] Add kibana dashboard(probably similar to what we have for docker)
- [ ] I added [TestData](https://github.com/kaiyan-sheng/beats/blob/fargate_module/x-pack/metricbeat/module/awsfargate/task_stats/task_stats_integration_test.go) function to create data.json file from the [mock response body](https://github.com/kaiyan-sheng/beats/tree/fargate_module/x-pack/metricbeat/module/awsfargate/task_stats/_meta/testdata) of task metadata endpoint. Is this OK? Do we have a function already to create `data.json` from mock data under `metricbeat/mb/testing`?
- [ ] I copied several functions from `docker` module, maybe we can refactor them in a better way so I don't have to duplicate the code.

File: metricbeat/docs/modules/docker/container.asciidoc [hunk 2]
```diff
@@ -7,7 +7,6 @@ This file is generated! See scripts/mage/docs_collector.go
 
 include::../../../module/docker/container/_meta/docs.asciidoc[]
 
-This is a default metricset. If the host module is unconfigured, this metricset is enabled by default.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/docker/container.asciidoc:10]: Are these changes intended? Maybe to happen on different PR?

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/cloudformation.yml:1]: Is this valid?

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          - Effect: Allow
+            Principal:
+              Service: ecs-tasks.amazonaws.com
+            Action: sts:AssumeRole
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy
+      Policies:
+        - PolicyName: !Sub 'EcsTaskExecutionRole-${AWS::StackName}'
+          PolicyDocument:
+            Version: 2012-10-17
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/cloudformation.yml:32]: is this version valid?

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                Action:
+                  - secretsmanager:GetSecretValue
+                Resource:
+                  - arn:aws:secretsmanager:us-east-2:428152502467:secret:ELASTIC_CLOUD_ID-CcbUG2
+                  - arn:aws:secretsmanager:us-east-2:428152502467:secret:ELASTIC_CLOUD_AUTH-vMz67n
+  TaskDefinition:
+    Type: AWS::ECS::TaskDefinition
+    Properties:
+      Family: deployment-task-metricbeat
+      Cpu: 256
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/cloudformation.yml:44]: Hmm, cpu=256? Does this means 256 vCPUS 🤔 ?

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Parameters:
+  SubnetID:
+    Type: String
+Resources:
+  Cluster:
+    Type: AWS::ECS::Cluster
+    Properties:
+      ClusterName: test-cloudformation-fargate
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/cloudformation.yml:9]:
```suggestion
      ClusterName: metricbeat-cloudformation-fargate
```

**Review by ChrsMark — COMMENTED** [review]
looks good already! left some comments for now!

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 3 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:1]: Yes I got it from https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/format-version-structure.html :)

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 4 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:32]: Yes, this is from AWS [documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-policy.html) example:
```
Type: 'AWS::IAM::Policy'
Properties:
  PolicyName: CFNUsers
  PolicyDocument:
    Version: 2012-10-17
    Statement:
      - Effect: Allow
        Action:
          - 'cloudformation:Describe*'
          - 'cloudformation:List*'
          - 'cloudformation:Get*'
        Resource: '*'
  Groups:
    - !Ref CFNUserGroup
```

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 5 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:44]: Good question :) I searched the AWS [documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskdefinition.html) and here is the definition: 
256 meaning `0.25 vCPU`
`Cpu` field represents the number of cpu units used by the task. If you are using the Fargate launch type, this field is required and you must use one of the following values, which determines your range of valid values for the memory parameter:

| Valid Values  | Equivalent vCPU value | Memory Parameters |
| ------------- | ------------- | ------------- |
| 256  | .25 vCPU  | Available memory values: 512 (0.5 GB), 1024 (1 GB), 2048 (2 GB) |
| 512  | .5 vCPU  | Available memory values: 1024 (1 GB), 2048 (2 GB), 3072 (3 GB), 4096 (4 GB) |
| 1024  | 1 vCPU  | Available memory values: 2048 (2 GB), 3072 (3 GB), 4096 (4 GB), 5120 (5 GB), 6144 (6 GB), 7168 (7 GB), 8192 (8 GB) |
| 2048  | 2 vCPU  | Available memory values: Between 4096 (4 GB) and 16384 (16 GB) in increments of 1024 (1 GB) |
| 4096  | 4 vCPU  | Available memory values: Between 8192 (8 GB) and 30720 (30 GB) in increments of 1024 (1 GB) |

File: metricbeat/docs/modules/docker/container.asciidoc (hunk 2 above)
Comment (kaiyan-sheng) [reply metricbeat/docs/modules/docker/container.asciidoc]: Hmm not sure what happened there but I manually fixed it. Good catch! Thank you!

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 5 above)
Comment (ChrsMark) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:44]: I would say is fine! At k8s we set `0.1` https://github.com/elastic/beats/blob/0d5ef7b3ccdbd4c83d2a2df018b654164383ecee/deploy/kubernetes/metricbeat-kubernetes.yaml#L180

File: metricbeat/docs/modules/awsfargate.asciidoc [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+[float]
+== Introduction to AWS ECS and Fargate
+Amazon Elastic Container Service (Amazon ECS) is a highly scalable, fast,
+container management service that makes it easy to run, stop, and manage
+containers. ECS has two launch types that can define how compute resources will
+be managed: ECS EC2 and ECS Fargate.
+
+* *ECS EC2*
+
+ECS EC2 launches containers that run on EC2 instances. Users have to manage EC2
```
Comment (ChrsMark) [inline metricbeat/docs/modules/awsfargate.asciidoc:33]: We could add a note about how one can monitor this kind of containers. I suspect this could be done by using `docker` module directly installed on the EC2 instances but need to completely verify that (access to docker socket etc). wdyt?

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 3 above)
Comment (ChrsMark) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:1]: looks like ancient history but ok! 🤷🏼‍♂️

File: x-pack/metricbeat/module/awsfargate/cloudformation.yml (hunk 4 above)
Comment (ChrsMark) [reply x-pack/metricbeat/module/awsfargate/cloudformation.yml:32]: 👍🏼

File: x-pack/metricbeat/module/awsfargate/task_stats/_meta/docs.asciidoc [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+==== Store Elastic Cloud Credentials into AWS Secret Manager
+If users are using Elastic Cloud, it's recommended to store cloud id and cloud
+auth into AWS secret manager. Here are the AWS CLI example:
+
+Create secret ELASTIC_CLOUD_AUTH:
+----
+aws --region us-east-1 secretsmanager create-secret --name ELASTIC_CLOUD_AUTH --secret-string XXX
+----
+
+Create secret ELASTIC_CLOUD_AUTH:
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/_meta/docs.asciidoc:36]:
```suggestion
Create secret ELASTIC_CLOUD_ID:
```

File: x-pack/metricbeat/module/awsfargate/task_stats/cpu.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+			}
+			u.cpus = realCPUCount
+		}
+
+	}
+	return u.cpus
+}
+
+// SystemDelta calculates system delta.
+func (u *cpuUsage) SystemDelta() uint64 {
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/cpu.go:86]: Wondering if we could somehow avoid duplicating this with https://github.com/elastic/beats/blob/949fb147336e41447c0f842d3e28d9f1073e926d/metricbeat/module/docker/cpu/helper.go#L122. Maybe by using a base interface and apply some kind of extensions?

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+		logger:    logger,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	ecsURI, ok := os.LookupEnv("ECS_CONTAINER_METADATA_URI_V4")
+	if !ok {
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:82]: Shouldn't this check happen on init instead of every fetch?

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+		err := fmt.Errorf("lookup $ECS_CONTAINER_METADATA_URI_V4 failed")
+		m.logger.Error(err)
+		return err
+	}
+
+	taskStatsEndpoint := fmt.Sprintf("%s/task/stats", ecsURI)
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:88]: nit: I think we could define these two paths as consts in the very begginging of the file so as to have easier access to them. wdyt?

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+		err := fmt.Errorf("queryTaskMetadataEndpoints failed: %w", err)
+		m.logger.Error(err)
+		return err
+	}
+
+	eventsMapping(report, formattedStats)
+	return nil
+}
+
+func queryTaskMetadataEndpoints(taskStatsEndpoint string, taskEndpoint string) ([]Stats, error) {
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:101]: nit: maybe `taskStatsEndpoint` and `taskEndpoint could be metricset's fields and `queryTaskMetadataEndpoints` stand a method of `(m *MetricSet)`

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go (hunk 10 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:82]: Good point, I will move it so it only checks the endpoint once!

File: x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go (hunk 12 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/task_stats/task_stats.go:101]: Good point! Done!

File: x-pack/metricbeat/module/awsfargate/task_stats/cpu.go (hunk 9 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/task_stats/cpu.go:86]: Actually if we make `cpuUsage` public, then we can reuse all the methods here in awsfargate. Please see the latest commit 🙂

Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/awsfargate/task_stats/cpu.go]: Aha never mind, I fixed the doc issue with https://github.com/elastic/beats/pull/22034/commits/9e61b9079c54e0dd1ca97855cf7912dbc746162e#diff-1333975e4127602a5d3ef9c0758fb8eeb86370576a49681bdadf4335f3bf93bfR161

File: metricbeat/docs/modules/awsfargate.asciidoc (hunk 7 above)
Comment (kaiyan-sheng) [reply metricbeat/docs/modules/awsfargate.asciidoc:33]: Good point! I wonder if we should only put one or two sentences here for this because it is mainly focusing on fargate?

Comment (ChrsMark) [reply metricbeat/docs/modules/awsfargate.asciidoc:33]: Sounds good!

**Review by ChrsMark — APPROVED** [review]
Looks great!

**Comment by kaiyan-sheng** [comment]
Dashboard is added in a separate PR: https://github.com/elastic/beats/pull/22745
